hazelcast/src/main/java/com/hazelcast/query/impl/getters/JsonDataGetter.java

```diff
@@ -18,39 +18,47 @@
 
 import com.fasterxml.jackson.core.JsonFactory;
 import com.fasterxml.jackson.core.JsonParser;
+import com.hazelcast.internal.nio.BufferObjectDataInput;
+import com.hazelcast.internal.serialization.Data;
 import com.hazelcast.internal.serialization.InternalSerializationService;
-import com.hazelcast.internal.serialization.impl.DataInputNavigableJsonAdapter;
 import com.hazelcast.internal.serialization.impl.NavigableJsonInputAdapter;
-import com.hazelcast.internal.serialization.Data;
 
 import java.io.IOException;
 
 import static com.hazelcast.internal.serialization.impl.HeapData.HEAP_DATA_OVERHEAD;
 
 public final class JsonDataGetter extends AbstractJsonGetter {
 
-    private static final int UTF_CHARACTER_COUNT_FIELD_SIZE = 4;
+    private static final int UTF_BYTE_COUNT_FIELD_SIZE = 4;
 
-    private JsonFactory factory = new JsonFactory();
+    private final JsonFactory factory = new JsonFactory();
 
-    private InternalSerializationService ss;
+    private final InternalSerializationService ss;
 
-    JsonDataGetter(InternalSerializationService ss) {
+    public JsonDataGetter(InternalSerializationService ss) {
         super(null);
+
         this.ss = ss;
     }
 
     protected JsonParser createParser(Object obj) throws IOException {
         Data data = (Data) obj;
-        return factory.createParser(data.toByteArray(),
-                HEAP_DATA_OVERHEAD + UTF_CHARACTER_COUNT_FIELD_SIZE,
-                data.dataSize() - UTF_CHARACTER_COUNT_FIELD_SIZE);
+
+        return factory.createParser(
+            data.toByteArray(),
+            HEAP_DATA_OVERHEAD + UTF_BYTE_COUNT_FIELD_SIZE,
+            data.dataSize() - UTF_BYTE_COUNT_FIELD_SIZE
+        );
     }
 
     @Override
-    protected NavigableJsonInputAdapter annotate(Object object) {
+    protected NavigableJsonInputAdapter annotate(Object object) throws IOException {
         Data data = (Data) object;
-        return new DataInputNavigableJsonAdapter(ss.createObjectDataInput(data),
-                HEAP_DATA_OVERHEAD + UTF_CHARACTER_COUNT_FIELD_SIZE);
+
+        assert data.isJson();
+
+        try (BufferObjectDataInput input = ss.createObjectDataInput(data)) {
+            return new NavigableJsonInputAdapter(input.readUTF(), 0);
```

Comment:
This additional deserialization `Data` -> `String` affects the amount of memory allocated when querying on unindexed JSON attribute.

```
commit 905c8a50

Benchmark                                                                    Mode  Cnt        Score        Error   Units
JsonBenchmark.queryOnJsonIndexedAttribute                                   thrpt   10      147,944 ±      5,610   ops/s
JsonBenchmark.queryOnJsonIndexedAttribute:·gc.alloc.rate.norm               thrpt   10  8271544,542 ±    169,463    B/op

commit 320b492c

Benchmark                                                                    Mode  Cnt        Score       Error   Units
JsonBenchmark.queryOnJsonIndexedAttribute                                   thrpt   10      164,960 ±     3,288   ops/s
JsonBenchmark.queryOnJsonIndexedAttribute:·gc.alloc.rate.norm               thrpt   10  4403908,434 ±   151,382    B/op
```

(benchmark is [here](https://gist.github.com/vbekiaris/834494c98beb3b1556be1fa87a3daf38))

Can we keep the separate JSON input adapter optimized for `Data` (to avoid full deserialization)? This would help relieve memory pressure whenever attribute extraction needs to take place (eg querying unindexed attributes, ingesting large amount of JSON data when indexes are defined etc). wdyt?

Reply:
This is exactly what currently happens in the master branch (see `DataInputNavigableJsonAdapter`): we create and input stream over `Data` and read characters one by one. But since there is no standard way of reading UTF-8 characters sequentially from the byte stream in Java (at least I am not aware of any), we use `Bits#readUtf8Char`, which cannot handle certain UTF characters properly. 

So basically there are two main approaches:
1) Leave everything as is - save memory, but accept an exception in case of a 4-byte character
2) Use the approach from this PR - more memory, but handle everything well

Or maybe there is some Java API I am not aware of which allow for reading UTF8 character one by one from the given `DataInput`?

Reply:
@ahmetmircik @mmedenjak @ihsandemir @taburet  Guys, could you please provide your comments on @vbekiaris concern? I agree that the solution proposed in the PR is questionable. 

In the previous ticket (https://github.com/hazelcast/hazelcast/pull/15238) we removed usages of `Bits#*Utf8Char` methods from serialization logic, except for a single usage in JSON deserialization. However, in those modified places of serialization code, we typically expect a single string, which is normally small, so additional materialization of `String` object is not a big concern. When it comes to JSON, it might be a *very long string*, encoding complex object. So this new materialization which I introduced may have a really bad effect on performance, especially on queries.

I propose to do the following:
1) Reject this PR
2) If we a really concerned with the handling of 4-byte characters, improve our `Bits#readUtf8Char` routine to handle them. This way the performance will not be affected.
3) As this may require considerable efforts on testing, the change is not breaking, and there is no indication that it affects any user at the moment, move the ticket to 4.1+ milestone.

What do you think?

Reply:
@devozerov I have been working on a solution using a `CharsetDecoder` supplied by `Bits.UTF_8.newDecoder()`. Its behaviour looks good in terms of managing multibyte and malformed sequences. There are still some details to iron out, I will post more soon.

Reply:
Performance was a big concern during the development of JSON querying (https://hazelcast.atlassian.net/wiki/spaces/PM/pages/907345951/JSON+Non-Indexed+Query+Optimisation - it even has its' own requirement and we had extensive benchmarking during TDD verification just to get it signed off) so I'm leaning towards either not fixing it OR some other approach that Vassilis might have in mind.

Reply:
Should it be part of 4.0 then? At the moment we do not handle 4-byte characters and an exception is thrown. If we fix it after 4.0 this should not cause any compatibility issues.

Reply:
We can see if @vbekiaris manages to get his idea done in time for 4.0 so we don't need to think about this anymore.

Reply:
@mmedenjak @devozerov I prepared #16234 as an alternative, can you have a look?